File: android-native/app/src/main/java/com/example/kpkn/domain/training/PeriodizationEngine.kt
package com.example.kpkn.domain.training

import com.example.kpkn.data.models.MesocycleGoal
import kotlin.math.roundToInt

/**
 * Escala series, reps, %1RM y RPE según el objetivo del mesociclo (ACCUMULATION /
 * INTENSIFICATION / REALIZATION / DELOAD), aplicando también el volumeModifier
 * declarado por el bloque de protocolo. Es el equivalente "de bolsillo" a un
 * motor de periodización: no reemplaza autorregulación real, pero asegura que
 * el volumen y la intensidad varíen de forma consistente por fase.
 */
object PeriodizationEngine {

    data class SetPrescription(
        val sets: Int,
        val reps: Int,
        val percentageRM: Double,
        val rpe: Double,
    )

    private fun volumeGoalMultiplier(goal: MesocycleGoal): Double = when (goal) {
        MesocycleGoal.ACCUMULATION -> 1.15
        MesocycleGoal.INTENSIFICATION -> 1.0
        MesocycleGoal.REALIZATION -> 0.8
        MesocycleGoal.DELOAD -> 0.55
        MesocycleGoal.CUSTOM -> 1.0
    }

    private fun repsForGoal(baseReps: Int, goal: MesocycleGoal): Int = when (goal) {
        MesocycleGoal.ACCUMULATION -> baseReps + 2
        MesocycleGoal.INTENSIFICATION -> baseReps
        MesocycleGoal.REALIZATION -> baseReps - 2
        MesocycleGoal.DELOAD -> baseReps - 1
        MesocycleGoal.CUSTOM -> baseReps
    }.coerceIn(1, 20)

    private fun rpeForGoal(goal: MesocycleGoal): Double = when (goal) {
        MesocycleGoal.ACCUMULATION -> 6.5
        MesocycleGoal.INTENSIFICATION -> 8.0
        MesocycleGoal.REALIZATION -> 9.0
        MesocycleGoal.DELOAD -> 5.5
        MesocycleGoal.CUSTOM -> 7.5
    }

    /** Series ajustadas por objetivo del bloque y por el volumeModifier declarado en el protocolo. */
    fun scaleSets(baseSets: Int, goal: MesocycleGoal, volumeModifier: Double?): Int {
        val multiplier = volumeGoalMultiplier(goal) * (volumeModifier ?: 1.0)
        return (baseSets * multiplier).roundToInt().coerceIn(1, (baseSets * 3).coerceAtLeast(1))
    }

    /**
     * Ondulación mínima de %1RM dentro del bloque: sube gradualmente de
     * intensityMin (semana 1) a intensityMax (última semana del bloque).
     */
    fun percentageForWeek(intensityMin: Int, intensityMax: Int, weekNumber: Int, totalWeeksInBlock: Int): Double {
        if (totalWeeksInBlock <= 1) return (intensityMin + intensityMax) / 2.0
        val progress = ((weekNumber - 1).toDouble() / (totalWeeksInBlock - 1).toDouble()).coerceIn(0.0, 1.0)
        return intensityMin + (intensityMax - intensityMin) * progress
    }

    fun prescriptionFor(
        goal: MesocycleGoal,
        baseSets: Int,
        baseReps: Int,
        volumeModifier: Double?,
        intensityMin: Int,
        intensityMax: Int,
        weekNumber: Int,
        totalWeeksInBlock: Int,
        repScheme: List<Int>? = null,
    ): SetPrescription = SetPrescription(
        sets = scaleSets(baseSets, goal, volumeModifier),
        reps = repScheme
            ?.getOrNull((weekNumber - 1).coerceAtLeast(0))
            ?.coerceIn(1, 20)
            ?: repsForGoal(baseReps, goal),
        percentageRM = percentageForWeek(intensityMin, intensityMax, weekNumber, totalWeeksInBlock),
        rpe = rpeForGoal(goal),
    )
}

File: ios-native/KPKNFit/KPKNFit/Domain/Training/PeriodizationEngine.swift
import Foundation

/// Paridad mínima de `PeriodizationEngine.kt` para progresión por bloque.
public enum PeriodizationEngine {
    public struct SetPrescription {
        public let sets: Int
        public let reps: Int
        public let percentageRM: Double
        public let rpe: Double
    }

    private static func volumeGoalMultiplier(_ goal: MesocycleGoal) -> Double {
        switch goal {
        case .ACCUMULATION: return 1.15
        case .INTENSIFICATION: return 1.0
        case .REALIZATION: return 0.8
        case .DELOAD: return 0.55
        case .CUSTOM: return 1.0
        }
    }

    public static func scaleSets(baseSets: Int, goal: MesocycleGoal, volumeModifier: Double?) -> Int {
        let multiplier = volumeGoalMultiplier(goal) * (volumeModifier ?? 1.0)
        let scaled = Int((Double(baseSets) * multiplier).rounded())
        return max(1, min(scaled, max(baseSets * 3, 1)))
    }

    public static func percentageForWeek(
        intensityMin: Int,
        intensityMax: Int,
        weekNumber: Int,
        totalWeeksInBlock: Int
    ) -> Double {
        if totalWeeksInBlock <= 1 { return Double(intensityMin + intensityMax) / 2.0 }
        let progress = min(1.0, max(0.0, Double(weekNumber - 1) / Double(totalWeeksInBlock - 1)))
        return Double(intensityMin) + Double(intensityMax - intensityMin) * progress
    }

    public static func prescriptionFor(
        goal: MesocycleGoal,
        baseSets: Int,
        baseReps: Int,
        volumeModifier: Double?,
        intensityMin: Int,
        intensityMax: Int,
        weekNumber: Int,
        totalWeeksInBlock: Int
    ) -> SetPrescription {
        let reps: Int = {
            switch goal {
            case .ACCUMULATION: return min(20, max(1, baseReps + 2))
            case .INTENSIFICATION: return min(20, max(1, baseReps))
            case .REALIZATION: return min(20, max(1, baseReps - 2))
            case .DELOAD: return min(20, max(1, baseReps - 1))
            case .CUSTOM: return min(20, max(1, baseReps))
            }
        }()
        let rpe: Double = {
            switch goal {
            case .ACCUMULATION: return 6.5
            case .INTENSIFICATION: return 8.0
            case .REALIZATION: return 9.0
            case .DELOAD: return 5.5
            case .CUSTOM: return 7.5
            }
        }()
        return SetPrescription(
            sets: scaleSets(baseSets: baseSets, goal: goal, volumeModifier: volumeModifier),
            reps: reps,
            percentageRM: percentageForWeek(
                intensityMin: intensityMin,
                intensityMax: intensityMax,
                weekNumber: weekNumber,
                totalWeeksInBlock: totalWeeksInBlock
            ),
            rpe: rpe
        )
    }
}
